OpenRA.Mods.Common/Traits/Buildings/Building.cs

```diff
@@ -204,12 +203,13 @@ public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingNa
 					{
 						var unitsAtPos = world.ActorMap.GetActorsAt(pos).Where(a => a.IsInWorld
 							&& (a.Owner == p || (allyBuildEnabled && a.Owner.Stances[p] == Stance.Ally))
-							&& a.Info.HasTraitInfo<GivesBuildableAreaInfo>());
+							&& a.Info.HasTraitInfo<GivesBuildableAreaInfo>()
+							&& a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>().ValidTypes.Contains(AdjacentType));
```

Comment:
It would be better if this was pulled out to a `Func<Actor, bool>` and used `TraitInfoOrDefault` and a null check instead of `HasTraitInfo`.